tool_path_planner/include/tool_path_planner/eigen_value_edge_generator.h

```diff
@@ -0,0 +1,176 @@
+/**
+ * @author Jorge Nicho
+ * @file edge_path_generator.h
+ * @date Nov 9, 2019
+ * @copyright Copyright (c) 2019, Southwest Research Institute
+ *
+ * @par License
+ * Software License Agreement (Apache License)
+ * @par
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * @par
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef INCLUDE_TOOL_PATH_PLANNER_EIGEN_VALUE_EDGE_GENERATOR_H_
+#define INCLUDE_TOOL_PATH_PLANNER_EIGEN_VALUE_EDGE_GENERATOR_H_
+
+#include <boost/optional.hpp>
+#include <pcl/PolygonMesh.h>
+#include <shape_msgs/Mesh.h>
+#include <pcl/point_cloud.h>
+#include <pcl/point_types.h>
+#include <pcl/PointIndices.h>
+#include <pcl/octree/octree_search.h>
+#include <tool_path_planner/path_generator.h>
+
+namespace tool_path_planner
+{
+
+
+/**
+ * @brief The EigenValueEdgeGenerator class
+ *
+ * @details Currently this approache use the the geometric feature 'Surface Variation' to identify edges/contour which
+ * does not work well. Need to explore the other feature below to see if they produce better results. This feature is
+ * go at detecting creases but not the contour.
+ *
+ * The information was pulled from the paper "Contour detection in unstructured 3D point cloud" by Timo Hackel,
+ * Jan D. Wegner, Konrad Schindler.
+ *
+ * Geometric features based on eigen values:
+ *   - Sum                 = sigma1 + sigma2 + sigma3
+ *   - Omnivariance        = (sigma1 + sigma2 + sigma3)^(1/3)
+ *   - Eigenentropy        = -(sigma1*ln(sigma1) + sigma2*ln(sigma2) + sigma3*ln(sigma3))
+ *   - Anisotropy          = (sigma1 - sigma3) / sigma1
+ *   - Planarity           = (sigma2 - sigma3) / sigma1
+ *   - Linearity           = (sigma1 - sigma2) / sigma1
+ *   - Surface Variation   = sigma3 / (sigma1 + sigma2 + sigma3)
+ *   - Sphericity          = sigma3 /sigma1
+ *   - First Order Moment  = See Paper Eq. 2
+ *   - Line Feature        = See Paper Eq. 3
+ *   - Orientation Feature = See Paper Eq. 4
+ *   - Verticality         = See Paper
+ */
+class EigenValueEdgeGenerator : public PathGenerator
+{
+public:
+
+  struct Config
+  {
+    /**
+     * @ingroup  hsv values computation using the covariance matrix
+     * @{
+     */
+    double octree_res = 0.005;    /** @brief resolution for the octree */
+    double search_radius = 0.01;  /** @brief radious used for grouping */
```

Comment:
Perhaps one could default the search_radius to some value based on the point density. Simply find minimum bounding box, to get point volume, then point/sq_meter is a measure of the density. A fraction of this value as the search radius should provide a good starting point even if the data is provided in different units.